### PR TITLE
Update to support stable PyO3

### PR DESCRIPTION
**Pull request by omerbenamram**
The code works also with PyO3 0.12 + numpy 0.12, but not with PyO3 0.13.
The tests however require PyO3 0.11 + numpy 0.11 + nightly - because of `inline-python` dependency.

All tests are passing :)

This was formatted using latest `rustfmt`, so if there's any styling that needs to be applied we should try and add it to `rustfmt.toml`.

Let me know if there's anything else that needs to be addressed to merge this.

### CODE REVIEW

**Comment by de-vri-es** [comment]
Hey, thank you for the PR. Updating the dependencies has been on the todo-list for a while.

It's very hard to review now because of all the style changes though. You're right that we should include custom formatting in `rustfmt.toml`, but it shouldn't be done in this PR. Do you still have a version without the formatting changes?

**Comment by de-vri-es** [comment]
I've pushed a rustfmt.toml and .editorconfig to master, along with applying the formatting to the crate. I've also updated your commits based on that, but you appear to have unticked the box to allow me to push to your branch. So if you want to can update this PR using the `update-to-stable` branch from this repo, or you can manually update your branch.

**Comment by omerbenamram** [comment]
@de-vri-es Thanks!, done :) only the actual changes will be visible now. 

BTW I didn't know there's an option to allow upstream to push to my branch, I'll look for that.

File: src/from_numpy.rs [hunk 1]
```diff
@@ -57,17 +57,16 @@ pub struct UnalignedArrayError;
 /// This function creates a const slice that references data owned by Python.
 /// The user must ensure that the data is not modified through other pointers or references.
 #[allow(clippy::needless_lifetimes)]
-pub unsafe fn matrix_slice_from_numpy<'a, O, N, R, C>(
+pub unsafe fn matrix_slice_from_numpy<'a, N, R, C>(
 	_py: pyo3::Python,
-	input: O,
+	input: &'a PyAny,
```
Comment (de-vri-es) [inline src/from_numpy.rs:62]: Why did you change this argument? It seems like an unnecessary backwards incompatibility, and less convenient for the caller.

File: src/from_numpy.rs [hunk 2]
```diff
@@ -203,8 +200,9 @@ where
 		return Err(make_error());
 	}
 
+	let t2: u32 = N::DATA_TYPE.into_ffi_dtype() as u32;
```
Comment (de-vri-es) [inline src/from_numpy.rs]:
```suggestion
```

File: src/from_numpy.rs [hunk 3]
```diff
@@ -203,8 +200,9 @@ where
...
 	// Check the data type of the input array.
-	if npyffi::array::PY_ARRAY_API.PyArray_EquivTypenums((*(*array).descr).type_num, N::typenum_default()) != 1 {
+	if npyffi::array::PY_ARRAY_API.PyArray_EquivTypenums((*(*array).descr).type_num, t2 as i32) != 1 {
```
Comment (de-vri-es) [inline src/from_numpy.rs]:
```suggestion
	if npyffi::array::PY_ARRAY_API.PyArray_EquivTypenums((*(*array).descr).type_num, N::ffi_dtype() as u32 as i32) != 1 {
```

File: src/to_numpy.rs [hunk 4]
```diff
@@ -17,7 +17,7 @@ where
 	for r in 0..matrix.nrows() {
 		for c in 0..matrix.ncols() {
 			unsafe {
-				*array.uget_mut((r, c)) = matrix[(r, c)];
+				*array.uget_mut((r, c)) = matrix[(r, c)].clone();
```
Comment (de-vri-es) [inline src/to_numpy.rs]: I pretty sure this is not needed, since `nalgebra::Scalar` implies `Copy`.

File: tests/from_numpy.rs [hunk 5]
```diff
@@ -11,201 +11,159 @@ mod assert;
 
 /// Test conversion of a numpy array to a Matrix3<f64>.
 #[test]
-#[rustfmt::skip]
```
Comment (de-vri-es) [inline tests/from_numpy.rs]: I added these on purpose to prevent the matrices from being squashed to a single line. I'm guessing this is the result of automatic merge conflict resolving, but could you restore the formatting of the matrices in the tests?

File: Cargo.toml [hunk 6]
```diff
@@ -13,10 +13,10 @@ keywords = ["python", "numpy", "nalgebra", "matrix", "conversion"]
 categories = ["science"]
 
 [dependencies]
-nalgebra = "0.19.0"
-numpy = "0.7.0"
-pyo3 = "0.8.2"
+nalgebra = "0.24.1"
+numpy = "0.11"
+pyo3 = { version = "0.11" }
```
Comment (de-vri-es) [inline Cargo.toml]:
```suggestion
pyo3 = "0.11" 
```

File: src/from_numpy.rs [hunk 7]
```diff
@@ -1,7 +1,7 @@
 use nalgebra::base::{SliceStorage, SliceStorageMut};
 use nalgebra::{Dynamic, Matrix};
 use numpy::npyffi;
-use numpy::npyffi::objects::PyArrayObject;
+use numpy::{npyffi::objects::PyArrayObject, Element};
```
Comment (de-vri-es) [inline src/from_numpy.rs]: It looks like you're using `numpy::Element` throughout the file. I like the explicitness of that, but I think that makes this an unused import.

```suggestion
use numpy::{npyffi::objects::PyArrayObject};
```

**Review by de-vri-es — COMMENTED** [review]
Thanks for updating the PR. I left a few in-line comments, but overall it looks good. Thanks also for updating the tests to the  `inline-python` API!

Sadly there are still issues that prevent `inline-python` from updating to `pyo3` 0.13, so we're kinda stuck until that can be resolved.

File: tests/from_numpy.rs (hunk 5 above)
Comment (omerbenamram) [reply tests/from_numpy.rs]: Whoops, deficiently merge resolving. Reverted and restored matricies.

File: src/from_numpy.rs (hunk 1 above)
Comment (omerbenamram) [reply src/from_numpy.rs:62]: https://docs.rs/pyo3/0.11.1/pyo3/types/struct.PyDict.html#method.get_item now returns `&'a PyAny.` 
And `PyAny ` no longer implements `AsRef<PyAny>`, so this bound prevents the code from compiling.

File: src/to_numpy.rs (hunk 4 above)
Comment (omerbenamram) [reply src/to_numpy.rs]: image

But AFAIK clone is specifically optimized for scalars - https://docs.rs/nalgebra/0.24.1/src/nalgebra/base/scalar.rs.html#26-28

**Comment by omerbenamram** [comment]
Thanks for the thorough CR. I've addressed the comments.

The reason I've settled with 0.11 is mostly because it's the first version to fully support stable rust.
I've looked into the 0.13 issues, and it appears there is indeed work to be done to support it both in the main code and in the tests, but IMO it could be done separately from this change set.

WDYT?

File: src/to_numpy.rs (hunk 4 above)
Comment (de-vri-es) [reply src/to_numpy.rs]: Ah, they dropped the `Copy` from `Scalar`. Fair enough. The trait documentation seems to suggest we should use `inlinded_clone()` to opt in to the inlined version in debug builds too. Could you still update that?

File: src/from_numpy.rs (hunk 1 above)
Comment (de-vri-es) [reply src/from_numpy.rs:62]: Ah, that is a shame. But fair enough, then this seems best for now.

File: src/to_numpy.rs (hunk 4 above)
Comment (de-vri-es) [reply src/to_numpy.rs]: Actually, let's just merge. I'll push a separate commit for that :)

Comment (omerbenamram) [reply src/to_numpy.rs]: done.

**Comment by de-vri-es** [comment]
> Thanks for the thorough CR. I've addressed the comments.

Thank you!
 
> The reason I've settled with 0.11 is mostly because it's the first version to fully support stable rust.
> I've looked into the 0.13 issues, and it appears there is indeed work to be done to support it both in the main code and in the tests, but IMO it could be done separately from this change set.

Yeah, that's good for now. It's already a big improvement :)

> WDYT?

Ready to merge :) I'll try to get a release out today too.

Thanks again for picking this up!

File: src/to_numpy.rs (hunk 4 above)
Comment (de-vri-es) [reply src/to_numpy.rs]: Ah, you beat me to it. Even better ;)

**Comment by de-vri-es** [comment]
Released as 0.3.0 :)
